### PR TITLE
Shop page with basic search

### PR DESCRIPTION
**Pull request by EmirKapic**

### CODE REVIEW

File: frontend/src/models/CategoryWithSubs.ts [hunk 1]
```diff
@@ -0,0 +1,9 @@
+import Subcategory from "./Subcategory";
+
+type CategoryWithSubs = {
```
Comment (EmirKapic) [inline frontend/src/models/CategoryWithSubs.ts:3]: Perhaps a bad name?

File: backend/src/main/java/com/atlantbh/internship/AuctionApp/dtoconvertors/CategoryDtoConverter.java
Comment (farispoljcic-abh) [inline backend/src/main/java/com/atlantbh/internship/AuctionApp/dtoconvertors/CategoryDtoConverter.java:1]: Not sure we actually need the `dtoconvertors` package. We could just add a static method inside `CategoryDto`:
```
public static CategoryDto fromModel(Category m) {
    return new CategoryDto(m.getId(), m.getName(), m.getSubCategories());
}
```
Though I'm not exactly sure why do we need `CategoryDto` at all, can we just include the subcategories whenever we fetch the category as well? Didn't you mention that for cases where we fetch a subcategory, the subcategories in the corresponding category would be an empty array?

We will inevitably need some DTOs, but DTOs that are the exact same as the model are way too confusing.

File: backend/src/main/java/com/atlantbh/internship/AuctionApp/models/SubCategory.java [hunk 2]
```diff
@@ -18,4 +21,6 @@ public class SubCategory {
     @JoinColumn(name = "category_id")
     private Category category;
 
+    private Long productCount;
```
Comment (farispoljcic-abh) [inline backend/src/main/java/com/atlantbh/internship/AuctionApp/models/SubCategory.java:24]: Can this be `null`?

File: backend/src/main/java/com/atlantbh/internship/AuctionApp/repositories/ProductRepository.java [hunk 3]
```diff
@@ -5,16 +5,18 @@
...
     @Query(value = "FROM Product where dateStart < current date and dateEnd > current date ORDER BY RANDOM() LIMIT 1")
     Product getRandom();
 
-    /**
-     * Both parameters should be set to Instant.now(). It must be done this way
-     * because spring jpa wont allow passing Instant.now() directly
-     */
-    Page<Product> findAllByDateEndAfterAndDateStartBefore(Pageable pageable, Instant today1, Instant today2);
+
+    @Query("from Product where (:categoryId is null or :categoryId = subCategory.category.id ) and (:subcategoryId is null or :subcategoryId = subCategory.id) and(:name is null or name like %:name%) and (dateStart < current date and dateEnd > current date )")
```
Comment (farispoljcic-abh) [inline backend/src/main/java/com/atlantbh/internship/AuctionApp/repositories/ProductRepository.java:17]: Use the java text block to format this a bit better.

File: backend/src/main/java/com/atlantbh/internship/AuctionApp/services/Product/ProductParameters.java [hunk 4]
```diff
@@ -0,0 +1,5 @@
+package com.atlantbh.internship.AuctionApp.services.Product;
+
+public record ProductParameters(Long categoryId, Long subcategoryId, String name){
+
+}
```
Comment (farispoljcic-abh) [inline backend/src/main/java/com/atlantbh/internship/AuctionApp/services/Product/ProductParameters.java]: Add newline at EOF

File: backend/src/main/java/com/atlantbh/internship/AuctionApp/services/Product/ProductServiceImpl.java [hunk 5]
```diff
@@ -8,16 +8,19 @@
...
 public class ProductServiceImpl implements ProductService {
     private ProductRepository productRepository;
 
     @Override
-    public Page<Product> getAllActive(Pageable pageable) {
-        return productRepository.findAllByDateEndAfterAndDateStartBefore(pageable, Instant.now(), Instant.now());
+    public Page<Product> getAllActive(Pageable pageable, ProductParameters params) {
+        if (params == null) {
+            params = new ProductParameters(null, null, null);
+        }
```
Comment (farispoljcic-abh) [inline backend/src/main/java/com/atlantbh/internship/AuctionApp/services/Product/ProductServiceImpl.java:22]: Aren't these parameters `null` by default?

File: frontend/src/components/Common/Button.tsx [hunk 6]
```diff
@@ -12,7 +12,9 @@ export interface ButtonProps {
 const buttonClassName: Record<ButtonType, string> = {
   primary:
     "border-2 border-purple flex gap-2 items-center justify-center duration-300 hover:bg-purple hover:text-white shadow-purple-md",
-  secondary: "...", //will be added when needed
+  secondary: "...",
```
Comment (farispoljcic-abh) [inline frontend/src/components/Common/Button.tsx:15]: Let's remove the `secondary` type until it's used.

File: frontend/src/components/Navbar/LowerNavbar.tsx [hunk 7]
```diff
@@ -1,9 +1,33 @@
...
 import Icon from "svgs/Icon";
 
 export default function LowerNavbar() {
   const { pathname } = useLocation();
+  const navigate = useNavigate();
+  const [searchText, setSearchText] = useState("");
+
+  const pathPrefix = pathname === "/" ? "/" : pathname.split("/")[1];
+
+  function handleSearch(e: FormEvent): void {
```
Comment (farispoljcic-abh) [inline frontend/src/components/Navbar/LowerNavbar.tsx:13]: Can we call this on enter as well, not just on submit button.

File: frontend/src/pages/Shop/Sidebar/CategoryListItem.tsx [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+    ]);
+    navigate(`/shop?${queryParams}`, {
+      state: {
+        pageReset: true,
+      },
+    });
+  }
+
+  const subCategories = props.category.subCategories
+    .sort((subCatA, subCatB) => subCatB.productCount - subCatA.productCount)
```
Comment (farispoljcic-abh) [inline frontend/src/pages/Shop/Sidebar/CategoryListItem.tsx:27]: This sort should be on backend.

File: frontend/src/pages/Shop/Sidebar/CategoryListItem.tsx [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+
+export interface CategoryListItemProps {
+  category: CategoryWithSubs;
+  active: boolean;
+}
+
+export default function CategoryListItem(props: CategoryListItemProps) {
+  const navigate = useNavigate();
+
+  function handleSelect(categoryId: string, subcategoryId?: string) {
```
Comment (farispoljcic-abh) [inline frontend/src/pages/Shop/Sidebar/CategoryListItem.tsx:14]: `: void`

File: frontend/src/pages/Shop/Sidebar/CategoryListItem.tsx [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+        <h1 className="text-sm">{props.category.name}</h1>
+        {props.active ? (
+          <button>
+            <Icon name="minus" />
+          </button>
+        ) : (
+          <button>
+            <Icon name="plus" />
+          </button>
+        )}
```
Comment (farispoljcic-abh) [inline frontend/src/pages/Shop/Sidebar/CategoryListItem.tsx:54]: Or just:
```
<button>
  <Icon name={props.active ? "minus" : "plus"} />
</button>
```

File: frontend/src/svgs/Icon.tsx [hunk 11]
```diff
@@ -83,5 +85,35 @@ export default function Icon(props: IconProps) {
...
+          strokeWidth={1.5}
+          stroke="currentColor"
+          className="w-6 h-6"
+        >
+          <path
+            strokeLinecap="round"
+            strokeLinejoin="round"
+            d="M12 4.5v15m7.5-7.5h-15"
+          />
+        </svg>
```
Comment (farispoljcic-abh) [inline frontend/src/svgs/Icon.tsx:103]: Out of curiosity, where do you find these SVGs?

Comment (EmirKapic) [reply frontend/src/svgs/Icon.tsx:103]: https://heroicons.com/

File: backend/src/main/java/com/atlantbh/internship/AuctionApp/dtos/CategoryDto.java [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+
+/**
+ * This dto is used because at some endpoints we need category with its subcategories,
+ * and at some we need subcategory inside of which is its category (like with ones connected to product)
+ * because of this we need to keep subCategories out of JSON response in some endpoints, hence the need for a specific dto
+ * when we DO need them
+ */
+@AllArgsConstructor
+@Data
+public class CategoryDto {
```
Comment (farispoljcic-abh) [inline backend/src/main/java/com/atlantbh/internship/AuctionApp/dtos/CategoryDto.java:19]: Could you change this to a record as well?

File: backend/src/main/java/com/atlantbh/internship/AuctionApp/repositories/ProductRepository.java [hunk 13]
```diff
@@ -5,16 +5,20 @@
...
-    /**
-     * Both parameters should be set to Instant.now(). It must be done this way
-     * because spring jpa wont allow passing Instant.now() directly
-     */
-    Page<Product> findAllByDateEndAfterAndDateStartBefore(Pageable pageable, Instant today1, Instant today2);
+
+    @Query("""
+            from Product where (:categoryId is null or :categoryId = subCategory.category.id )
+            and (:subcategoryId is null or :subcategoryId = subCategory.id)
+            and(:name is null or name like %:name%)
```
Comment (farispoljcic-abh) [inline backend/src/main/java/com/atlantbh/internship/AuctionApp/repositories/ProductRepository.java:18]: We should ignore upper and lower case differences when searching by the name.

File: frontend/src/models/CategoryWithSubs.ts (hunk 1 above)
Comment (farispoljcic-abh) [inline frontend/src/models/CategoryWithSubs.ts:3]: Just use the same name as it is on backend, `CategoryDto`

**Review by farispoljcic-abh — APPROVED** [review]

**Pull request merged**